### ISSUE TITLE
Book uses academic term "arity" without definition

### ISSUE THREAD

**Issue by snej** [rust-lang/rust#21568]
In the [Compound Data Types](http://doc.rust-lang.org/book/compound-data-types.html) chapter of the book, the term "arity" is used without any prior definition:

>You can assign one tuple into another, if they have the same arity and contained types.

I've learned enough formal logic and programming language theory to know what "arity" means, but I suspect most programmers don't. (Neither does my spell checker!) And the book otherwise seems to be trying to use clear language and be readable by non-experts. I suggest changing it to:

>You can assign one tuple into another, if they have the same number and type of items.

**Comment by kmcallister** [comment]
Sounds great to me, feel free to submit a pull request :)  Maybe with "types" instead of "type".

**Comment by kmcallister** [comment]
Or:

> You can assign one tuple into another, if the sequence of types in each is exactly the same.

**Comment by steveklabnik** [comment]
I'd prefer just defining `arity`. it's a really useful word.

**Comment by snej** [comment]
It's a useful word if you do a lot of functional programming or formal logic ;-), but the rest of the book is very informal and tries hard to make J. Random Scripter feel comfortable, so tossing in a $20 word seems out of place.

Maybe:

>…if the sequence of types (their [_arity_](https://en.wikipedia.org/wiki/Arity)) is exactly the same.

**Comment by kmcallister** [comment]
But the arity refers to just the number of entries, not their types. It's an incomplete description and so awkward to define in passing.

**Comment by Gankro** [comment]
I don't think arity need be mentionned. Saying "contain the same types" seems sufficient. That already has to deal with (A, B) vs (B, A). Seems to me that also deals with (A, A) vs (A, A, A).

**Comment by hartsantler** [comment]
I had seen `arity` in the docs to, and it simply pissed me off.  Why force me to use google and dictionary.com?  Please just explain things in plain english, the docs is not a place to show off how smart you are and how advanced your english is.

**Comment by steveklabnik** [comment]
@hartsantler it's not to show off fancy language, it's a computer science term. It's a good word for any programmer to know.

**Comment by hartsantler** [comment]
@steveklabnik sometimes plain english is simply better, these are docs are supposed to make it easy to learn rust.  The reader is not likely interested in a course on rarely used computer science terms.  Just have a look part of the definition of `arity` on wikpedia.  Cartesian product? dyadic? valency? adicity and degree?

This term `arity` comes off as very elitist.  Save it for your own blog, not the docs.

"In logic, mathematics, and computer science, the arity Listeni/ˈærɨti/ of a function or operation is the number of arguments or operands the function or operation accepts. The arity of a relation (or predicate) is the dimension of the domain in the corresponding Cartesian product. (A function of arity n thus has arity n+1 considered as a relation.) The term springs from words like unary, binary, ternary, etc. Unary functions or predicates may be also called "monadic"; similarly, binary functions may be called "dyadic".
In mathematics arity may also be named rank,[1][2] but this word can have many other meanings in mathematics. In logic and philosophy, arity is also called adicity and degree.[3][4] In linguistics, arity is usually named valency.[5]"

**Comment by emanueLczirai** [comment]
:+1: for keeping `arity` somehow. It's well worth learning the word imho.

**Comment by kmcallister** [comment]
@hartsantler: This attitude is not going to serve you well in Rust.  It's not "elitist" to define precise terms for concepts that may be hard to express in "plain english".

Of course one *should* define them in the text.  And in this case I don't think it's worth the trouble, because there *are* satisfactory explanations in "plain english" (see above).  So, I'm in agreement with you that the text needs to be changed.  But turning it into a rant about "elites" is just gross and unhelpful.

The thing is... when people ignore your ideas while making fun of the words you use to express them, it really *does* breed a bitterness and elitism, and ultimately a sort of siege mentality that you'll see in a lot of functional programming communities.  It's not pretty, and I don't want Rust to go the same way, although I don't think it's particularly likely.

So please let's leave the culture war out of it.  Technical writing is *hard*. There are always many tradeoffs.  Just because someone uses a word you don't know, doesn't mean they're maliciously trying to feel superior to you.  My unsolicited advice: Embrace being around people who know things you don't. Then you can learn from them. I hope that the Rust community will *always* be a safe place for that.

**Comment by mdinger** [comment]
Slight modification. I think *signature* is the right word. Even if signatures' definition is unclear, the example should make it's meaning understandable; something the *arity* lacked.

---

You can assign tuples onto each other if the signature matches:
```rust
let mut x = (1i32, 2i64);
let y =     (2i32, 3i64);
let fail =  (2i64, 5i32);

x = y;    // Works because signature is the same
x = fail; // Correct types, wrong order. This doesn't work.
```

**Comment by jxcl** [comment]
@steveklabnik The PR that addresses this ticket has been merged. Should this issue be closed now?

**Comment by steveklabnik** [comment]
Yes, thanks.
